### PR TITLE
Fix issue #33242: ExtractMethodCodeRefactoringProvider throws invalid cast when used within the body of a constructor expression-member body

### LINKED ISSUE #33242: ExtractMethodCodeRefactoringProvider throws invalid cast when used within the body of a constructor expression-member body
1. Highlight between the $$ (after removing them)

``` C#

        private sealed class GraphIconCache
        {
            private ImmutableHashSet<ImageMoniker> _registeredIcons = ImmutableHashSet<ImageMoniker>.Empty;

            private ImmutableDictionary<(int id, Guid guid), string> _iconNameCache = ImmutableDictionary<(int id, Guid guid), string>.Empty;

            private readonly IVsImageService2 _imageService;

            public static async Task<GraphIconCache> CreateAsync(IAsyncServiceProvider serviceProvider)
            {
                var imageService = (IVsImageService2)await serviceProvider.GetServiceAsync(typeof(SVsImageService));

                return new GraphIconCache(imageService);
            }

            private GraphIconCache(IVsImageService2 imageService) => _imageService = $$imageService$$;

            public string GetName(ImageMoniker icon)
            {
                return ImmutableInterlocked.GetOrAdd(ref _iconNameCache, (id: icon.Id, guid: icon.Guid), i => $"{i.guid:D};{i.id}");
            }

            public void Register(ImageMoniker icon)
            {
                if (ImmutableInterlocked.Update(ref _registeredIcons, (knownIcons, arg) => knownIcons.Add(arg), icon))
                {
                    _imageService.TryAssociateNameWithMoniker(GetName(icon), icon);
                }
            }
        }

```
```
System.InvalidCastException : Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.ConstructorDeclarationSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax'.
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.CodeGenerator`3.AddInvocationAtCallSiteAsync[TStatement,TExpression,TNodeUnderContainer](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CSharp.ExtractMethod.CSharpMethodExtractor.CSharpCodeGenerator.CreateStatementsOrInitializerToInsertAtCallSiteAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CSharp.ExtractMethod.CSharpMethodExtractor.CSharpCodeGenerator.GenerateBodyForCallSiteContainerAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.CodeGenerator`3.GenerateAsync[TStatement,TExpression,TNodeUnderContainer](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.MethodExtractor.ExtractMethodAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.ExtractMethod.AbstractExtractMethodService`3.ExtractMethodAsync[TValidator,TExtractor,TResult](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.GetCodeActionAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.ExtractMethod.ExtractMethodCodeRefactoringProvider.ComputeRefactoringsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringService.GetRefactoringFromProviderAsync(<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)

```

### PR DESCRIPTION
**Pull request by tiesmaster**
Fixes #33242 
This crash was caused by [`CSharpCodeGenerator.CreateStatementsOrInitializerToInsertAtCallSiteAsync()` ](https://github.com/dotnet/roslyn/blob/de57f6daeee86908cc9dcd0386d2f66d60e8cee1/src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs#L137-L140) not evaluating the expression bodied constructor as such, since [`MemberDeclarationSyntaxExtensions.GetExpressionBody()`](https://github.com/dotnet/roslyn/blob/de57f6daeee86908cc9dcd0386d2f66d60e8cee1/src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs#L282-L299) wasn't updated yet to handle expression bodied constructors.

It seemed most appropriate to handle that in `MemberDeclarationSyntaxExtensions`, and not in `CSharpMethodExtractor.CSharpCodeGenerator`.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs [hunk 1]
```diff
@@ -293,6 +293,8 @@ public static ArrowExpressionClauseSyntax GetExpressionBody(this MemberDeclarati
                     return ((OperatorDeclarationSyntax)memberDeclaration).ExpressionBody;
                 case SyntaxKind.ConversionOperatorDeclaration:
                     return ((ConversionOperatorDeclarationSyntax)memberDeclaration).ExpressionBody;
+                case SyntaxKind.ConstructorDeclaration:
+                    return ((ConstructorDeclarationSyntax)memberDeclaration).ExpressionBody;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: are destructors something we need to handle as well?

Comment (tiesmaster) [reply src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: Good point. I'd say so. I'll add another test case, and handle that.

Comment (tiesmaster) [reply src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: BTW I also noticed that this entire `switch` can be condensed into the following, however, from the other code in that file, this doesn't looks like something that gets done. Is this kind of refactoring appreciated/encouraged, or rather frown upon?

````csharp
switch (memberDeclaration.Kind())
{
    case SyntaxKind.MethodDeclaration:
    case SyntaxKind.OperatorDeclaration:
    case SyntaxKind.ConversionOperatorDeclaration:
    case SyntaxKind.ConstructorDeclaration:
    case SyntaxKind.DestructorDeclaration:
        return true;
}
````

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: > Is this kind of refactoring appreciated/encouraged, or rather frown upon?

I would have to see the change suggested.  Based on that, i could tell you.  Worst case, you simply revert that change in the PR. :)

Comment (tiesmaster) [reply src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: Ah, ok 😅 Well, this then: [9262ad4](https://github.com/dotnet/roslyn/pull/33462/commits/9262ad4076c5f7c85fee42f94fe2a5a6f2b09431).

Please have another look.

File: src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs [hunk 2]
```diff
@@ -284,19 +284,13 @@ public static ArrowExpressionClauseSyntax GetExpressionBody(this MemberDeclarati
...
-                    return ((IndexerDeclarationSyntax)memberDeclaration).ExpressionBody;
                 case SyntaxKind.OperatorDeclaration:
-                    return ((OperatorDeclarationSyntax)memberDeclaration).ExpressionBody;
                 case SyntaxKind.ConversionOperatorDeclaration:
-                    return ((ConversionOperatorDeclarationSyntax)memberDeclaration).ExpressionBody;
                 case SyntaxKind.ConstructorDeclaration:
-                    return ((ConstructorDeclarationSyntax)memberDeclaration).ExpressionBody;
                 case SyntaxKind.DestructorDeclaration:
-                    return ((DestructorDeclarationSyntax)memberDeclaration).ExpressionBody;
+                    return ((BaseMethodDeclarationSyntax)memberDeclaration).ExpressionBody;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: this won't work (afaict).  For example, IndexerDeclaration and PropertyDeclaration do not derive from BaseMethodDeclarationSyntax.

Comment (tiesmaster) [reply src/Workspaces/CSharp/Portable/Extensions/MemberDeclarationSyntaxExtensions.cs]: Ah, good point. I checked most of the nodes, and tested it locally, and it passed, though, I guess there isn't coverage for those. I've reverted, don't think it adds that much value, in this case to refactor, especially if it requires adding a lot of "synthetic" coverage.

**Pull request merged**